recipe/2024-04-terraform-provider-secrets.md

```diff
@@ -0,0 +1,259 @@
+
+### High Level Design
+
+The design for handling secrets for Private Terraform modules established the basis for data flow between Engine and Driver components in the codebase. For Terraform Providers, we will continue to maintain the flow where secrets are retrieved in the Engine and passed into the Driver in the Execute() call. 
+The following diagram describes the updated data flow. The calls in green highlight updates to existing code.
+
+
+![alt text](2024-04-terraform-provider-secrets.png)
+
+
+### Detailed Design
+
+The FindSecretIds function in the DriverWithSecrets interface is updated to return a struct of RecipeSecretIds, which will hold secret Ids for a Terraform module as before and in addition, also populate secret Ids for provider configuration and environment variables, iterating through the input environment recipe configuration. 
+
+The LoadSecrets method in the SecretLoader interface will also be updated to LoadRecipeSecrets and return struct RecipeSecrets which will contain secrets for module, providers and environment variables.
+
+#### Advantages (of each option considered)
+Retrieving secrets in the Engine and passing them into the Driver was discussed positively as a way forward. Looking ahead with Containerization, the Driver will have all the information it needs to be able to be spin off into a separate process.
+
+#### Disadvantages (of each option considered)
+N/A
+
+#### Proposed Option
+N/A
+
+### API design (if applicable)
+N/A
+
+### CLI Design (if applicable)
+N/A
+
+### Implementation Details
+####Recipes (Engine and Driver)
+
+```
+Note: DataModel in [Design Document to support multiple Terraform Providers](https://github.com/radius-project/design-notes/pull/39/files)
+DataModel in Design doc describes SecretReference:
+type SecretReference {
+  source: string;		// SecretStore Id					
+  key: string;
+}
+
+type RecipeSecretIDs struct {
+  moduleSecret string                  // SecretStoreId associated with git private terraform repo source
+  providerSecrets map[string]map[string]SecretReference  // map[providerName or provider.aliasName][secret_name]SecretReference
+  envSecrets map[string]SecretReference // map[environmentvariablename]SecretReference 
+}
+
+type RecipeSecrets struct {
+  moduleSecrets v20231001preview.SecretStoresClientListSecretsResponse
+  providerSecrets map[string]map[string]string   // map[providerName/provider.aliasName][secret_name]value
+  envSecrets map[string]string                   // map[environmentvariablename]value 
+}
+
+
+// FindSecretIDs will be updated in current DriverWithSecrets interface
+type DriverWithSecrets interface {
+	// Driver is an interface to implement recipe deployment and recipe resources deletion.
+	Driver
+
+	// FindSecretIDs gets the secret store resource IDs for the recipe including module, provider, envSecrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (**RecipeSecretIDs**, error)
```

Comment:
What would we lose if we returned `([]string, error)` from this function?

Reply:
updated this currently to map[string]string. Also updated 'Alternatives Considered' section to describe we could go with []string.